File: LeetCode/symmetric_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
public:
    bool solve(TreeNode* r1, TreeNode* r2) {
        if(!r1 && !r2)
            return true;
        
        if((!r1 && r2) || (r1 && !r2))
            return false;
        
        if(r1->val != r2->val)
            return false;
        
        return solve(r1->left, r2->right) && solve(r1->right, r2->left);
    }
    
    bool isSymmetric(TreeNode* root) {
        return solve(root, root);
    }
};

File: LeetCode/symmetric_tree.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func solve(root1 *TreeNode, root2 *TreeNode) bool {
    if root1 == nil && root2 == nil {
        return true
    } 

    if root1 == nil && root2 != nil {
        return false
    }

    if root1 != nil && root2 == nil {
        return false
    }

    return root1.Val == root2.Val && solve(root1.Left, root2.Right) && solve(root1.Right, root2.Left)
}

func isSymmetric(root *TreeNode) bool {
    return solve(root, root)
}
